docs/DesignDocs/CustomAttestation.md

```diff
@@ -0,0 +1,420 @@
+or a timestamp. Another user might want their enclaves to generate attestation
+data that is in a compatible format with their existing authentication
+infrastructure, such as a JSON Web Token or a X509 certificate. Overall,
+there has been interest in enchancing Open Enclave's APIs to support custom
+attestation formats to enable these scenarios.
+
+Specification
+-------------
+
+To support custom attestation formats, this document proposes adding a plugin
+model for attestaton. The Open Enclave SDK will define a set of common APIs
+that each plugin must implement. Each plugin will define a random UUID to
+distinguish it from other plugins.
+
+Futhermore, there will be additional attestation "plugin aware" APIs that are
+analogous to `oe_get_report` and `oe_verify_report`, along with functions for
+registering and unregistering plugins. The user can statically link in their
+desired plugin and call the register plugin function. The attestation data can
+be retrieved from the "plugin aware" analogue of `oe_get_report` with the
+desired UUID. The generated data will have the UUID in its header. The user can
+call the analogue of `oe_verify_report` to verify the data and the Open Enclave
+runtime can use this UUID to determine what plugin verification routine to run.
+
+### Plugin API
+
+Each plugin must implement the functions below:
+
+```C
+/**
+ * Struct that defines the structure of each plugin. Each plugin must
+ * define an UUID for its format and implement the functions in this
+ * struct. Ideally, each plugin should provide a helper function to
+ * create this struct on the behalf of the plugin users.
+ */
+struct oe_attestaton_plugin_t
+{
+    /**
+     * The UUID for the plugin.
+     */
+    uuid_t format_id;
+
+    /**
+     * The function that gets executed when a plugin is registered.
+     *
+     * @param[in] plugin_context A pointer to the attestation plugin struct.
+     * @param[in] config_data An optional pointer to the configuration data.
+     * @param[in] config_data_size The size in bytes of config_data.
+     * @retval OE_OK on success.
+     */
+    oe_result_t (*on_register)(
+        oe_attestaton_plugin_t* plugin_context,
+        const void* config_data,
+        size_t config_data_size);
+
+    /**
+     * The function that gets executed when a plugin is unregistered.
+     *
+     * @param[in] plugin_context A pointer to the attestation plugin struct.
+     * @retval OE_OK on success.
+     */
+    oe_result_t (*on_unregister)(
+        oe_attestaton_plugin_t* plugin_context);
+
+    /**
+     * Generates the attestation evidence, which is defined as the data
+     * produced by the enclave. The caller may pass in custom claims, which
+     * must be attached to the evidence and then cryptographically signed.
+     *
+     * @param[in] plugin_context A pointer to the attestation plugin struct.
+     * @param[in] custom_claims A buffer to the optional custom clams.
```

Comment:
There are a few places where 'claims' is misspelled.

Reply:
Clams sound delicious! :-)

Reply:
hahaha